server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -151,6 +153,68 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         final long startTime = relativeTime();
         final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
 
+        boolean hasIndexRequestsWithPipelines = false;
+        final MetaData metaData = clusterService.state().getMetaData();
+        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
+        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
+            IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
+            if (indexRequest != null) {
+                // get pipeline from request
+                String pipeline = indexRequest.getPipeline();
+                if (pipeline == null) {
+                    // start to look for default pipeline via settings found in the index meta data
+                    IndexMetaData indexMetaData = indicesMetaData.get(actionRequest.index());
+                    if (indexMetaData == null && indexRequest.index() != null) {
+                        // if the write request if through an alias use the write index's meta data
+                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
+                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
+                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
+                            indexMetaData = alias.getWriteIndex();
+                        }
+                    }
+                    if (indexMetaData != null) {
+                        // Find the the default pipeline if one is defined from and existing index.
+                        String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexMetaData.getSettings());
+                        indexRequest.setPipeline(defaultPipeline);
+                        if (IngestService.NOOP_PIPELINE_NAME.equals(defaultPipeline) == false) {
+                            hasIndexRequestsWithPipelines = true;
+                        }
+                    } else if (indexRequest.index() != null) {
+                        // No index exists yet (and is valid request), so matching index templates to look for a default pipeline
+                        List<IndexTemplateMetaData> templates = MetaDataIndexTemplateService.findTemplates(metaData, indexRequest.index());
+                        assert (templates != null);
+                        String defaultPipeline = IngestService.NOOP_PIPELINE_NAME;
+                        // apply templates, here, in reverse order, since first ones are better matching
+                        for (int i = templates.size() - 1; i >= 0; i--) {
```

Comment:
This is a little confusing to me, shouldn't we just loop in standard order and break out once we have a pipeline?
We're doing the reverse iteration elsewhere because we merge all the template's settings but for just this one setting this seems wasteful?

Reply:
The template with the highest order should be used. The `MetaDataIndexTemplateService.findTemplates(...)` method orders the template by order in descending order. So we can just pick the first index template? (even though that template may not have a default pipeline defined)

Reply:
I need to emulate how the setting is retrieved for the concrete index so that the behavior here is the same on the first document as it as on the remainder.

For example, say that an index is manually created and matched 3 templates, a default pipeline is defined on the second template match. The index gets that set as the default pipeline. If we picked just the first matching template then we risk choosing a different (or none at all) default pipeline for the first document. 

It is parallel logic which isn't ideal, but feel that it relatively safe here. 

For example:
```
PUT _template/one
{
  "index_patterns": "foobar",
  "order": 1
}

PUT _template/two
{
  "index_patterns": "foobar",
  "order": 2,
  "settings": {
    "default_pipeline": "two"
  }
}

PUT _template/three
{
  "index_patterns": "foobar",
  "order": 3
}

PUT foobar
GET foobar
```
results in `"default_pipeline" : "two",` , which will be the pipeline used that we need to match for the very first document.

Reply:
But in your example, don't we just loop to the last template here and get settings that contain any default pipeline and set the default (because nothing is set) `IngestService.NOOP_PIPELINE_NAME` which would be wrong?

Reply:
@original-brownbear - good catch, i think you are right.  I am only test for precedence for templates with the `default_pipeline` defined. I believe that I can first check to see if the setting exists in the template thereby ignoring the default `IngestService.NOOP_PIPELINE_NAME`.  

Will update.

Reply:
@original-brownbear - updated https://github.com/elastic/elasticsearch/pull/39607/commits/efdaf538a47d7aa0cb289524541fd5adeaf871a9 (and updated test to cover this case)

Reply:
@jakelandis thanks, this looks like it will yield the correct result, but (I fully see the possibility that I"m wrong here :D), why not simply:

```java
for (IndexTemplateMetaData template : templates) {
                        final Settings settings = template.settings();
                        if (IndexSettings.DEFAULT_PIPELINE.exists(settings)) {
                            defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(settings);
                            break;
                        }
                    }
```
 
this seems to have the exact same effect as iterating in reverse order and simply overwriting a bunch of values until you iterated through the whole list?

Reply:
Thanks @jakelandis for explaining. I see now that simply picking the last index template is wrong.

I think that @original-brownbear suggestion make sense here. A normal loop based on the list returned by `MetaDataIndexTemplateService.findTemplates(...)` and break when the first pipeline is found is clearer then reserve iterating what the list returned by `MetaDataIndexTemplateService.findTemplates(...)`.

Reply:
@original-brownbear @martijnvg - thanks that does make sense for a single setting ! (I was bit slow understanding).

Changed loop here: https://github.com/elastic/elasticsearch/pull/39607/commits/80d71b43445535f47173481c4946491b8e50144e

Reply:
@jakelandis thanks LGTM :)